pinot-core/src/main/java/com/linkedin/pinot/core/util/IntDoubleIndexedPriorityQueue.java

```diff
@@ -0,0 +1,343 @@
+    _reusablePair.setDoubleValue(_values.get(0));
+    return _reusablePair;
+  }
+
+  /**
+   * Returns true if the priority queue is empty, false otherwise.
+   *
+   * @return True if empty, false otherwise
+   */
+  public boolean isEmpty() {
+    return _values.isEmpty();
+  }
+
+  /**
+   * Helper method that moves the element at the specified index up
+   * until the heap ordering is established.
+   *
+   * @param index Index of element to sift up.
+   */
+  private void siftUp(int index) {
+    // Return if already at root node.
+    if (index == 0) {
+      return;
+    }
+
+    while (index != 0) {
+      int parentIndex = getParentIndex(index);
+      double value = _values.get(index);
+      double parentValue = _values.get(parentIndex);
+
+      if (compare(parentValue, value) == 1) {
+        swapValues(index, parentIndex);
+        index = parentIndex;
+      } else {
+        // No more sifting up required, break
+        break;
+      }
+    }
+  }
+
+  /**
+   * Helper method that moves the element at the specified index down
+   * until the heap ordering is established.
+   *
+   * @param index Index of element to sift down.
+   * @return True if sifted, false otherwise.
+   */
+  private boolean siftDown(int index) {
+    boolean hasChildren = hasChildren(index);
+    if (!hasChildren) {
+      return false;
+    }
+
+    boolean sifted = false;
+    while (true) {
+      int leftChildIndex = getLeftChildIndex(index);
+      int rightChildIndex = getRightChildIndex(index);
+
+      int minIndex;
+      int size = _values.size();
+      if (leftChildIndex >= size && rightChildIndex >= size) { // This is leaf node, all done.
+        break;
+      } else if (rightChildIndex >= size) { // Node only has left child which will be the minimum.
+        minIndex = leftChildIndex;
+      } else { // Node has both left and right children, find the minimum of the two.
+        double leftChildValue = _values.get(leftChildIndex);
+        double rightChildValue = _values.get(rightChildIndex);
+
+        if (compare(leftChildValue, rightChildValue) <= 0) {
+          minIndex = leftChildIndex;
+        } else {
+          minIndex = rightChildIndex;
+        }
+      }
+
+      // One of the children is out of order, need to sift
+      if (compare(_values.get(index), _values.get(minIndex)) == 1) {
+        swapValues(index, minIndex);
+        index = minIndex;
+        sifted = true;
+      } else {
+        break;
+      }
+    }
+    return sifted;
+  }
+
+  /**
+   * Compares the two specified values, and returns:
+   * <ul>
+   *   <li> if v1 < v2, -1 for max, +1 for min mode. </li>
+   *   <li> if v1 > v2, -1 for max, -1 for min mode. </li>
+   *   <li> if v1 = v2,  0 for max, 0 for min mode. </li>
+   * </ul>
+   * @param v1 Value to compare
+   * @param v2 Value to compare
+   * @return Result of comparison (as described above).
+   */
+  private int compare(double v1, double v2) {
```

Comment:
Addressed all comments. Extracted out common code between the two PQ implementations into an abstract base class. While a lot of the code still 'seems' similar but it cannot be shared as one is based on primitive data types (for performance), while the other one is generic implementation.